hw30/js/index.js

```diff
@@ -0,0 +1,190 @@
+        this.#singleClickTaskIndex = this.#getTaskIndex(text);
+    }
+
+    #selectTaskByIndex(index) {
+        this.#tasks[index].selected = true;
+    }
+
+    clearSingleClickIndex() {
+        this.#singleClickTaskIndex = null;
+    }
+
+    addTaskPrepend() {
+        this.#addTask('prepend');
+    }
+
+    addTaskAppend() {
+        this.#addTask('append');
+    }
+
+    removeSelectedTasks() {
+        this.#tasks = this.#tasks.filter(task => !task.selected);
+    }
+
+    sortSelectedTasks() {
+        this.#tasks.sort((a, b) => b.selected - a.selected);
+    }
+
+    selectTask(text) {
+        const target = this.#tasks.find(task => task.text === text);
+
+        if (target) {
+            target.toggleSelected();
+        }
+    }
+
+    selectSingleTask(text) {
+        this.#clearSelected();
+        this.#setSingleClickIndex(text);
+        this.selectTask(text);
+    }
+
+    selectRange(text) {
+        this.#clearSelected();
+
+        const firstSelected = this.#getFirstSelectedTaskIndex();
+        const startIndex = this.#singleClickTaskIndex || (firstSelected >= 0 ? firstSelected : 0 );
+        const endIndex = this.#getTaskIndex(text);
+
+        if (startIndex < endIndex) {
+            for (let i = startIndex; i <= endIndex; i++) {
+                this.#selectTaskByIndex(i);
+            }
+        } else {
+            for (let i = startIndex; i >= endIndex; i--) {
+                this.#selectTaskByIndex(i);
+            }
+        }
+    }
+}
+
+class TodoApp {
+    #app = null;
+    #tasks = null;
+    #actions = null;
+    #content = null;
+    #list = null;
+
+    constructor(selector, initialTasks = []) {
+        this.#app = document.querySelector(selector);
+        this.#tasks = initialTasks;
+
+        this.#actions = this.#app.querySelector('.actions');
+        this.#content = this.#app.querySelector('.content');
+
+        this.#list = new TaskList(this.#tasks);
+
+        this.#render();
+        this.#addEvents()
+    }
+
+    #addEvents() {
+        this.#actions.addEventListener('click', this.#actionClick.bind(this));
+        this.#content.addEventListener('click', this.#listClick.bind(this));
+    }
+
+    #actionClick(event) {
+        const { target } = event;
+        const action = target.dataset['action'];
+
+        if (action) {
+            this.#list[action]();
+
+            this.#render();
+        }
+    }
+
+    #listClick(event) {
+        const { target, ctrlKey, metaKey, shiftKey } = event;
+        const { textContent } = target;
+
+        if (target.tagName !== 'LI') {
+            return;
+        }
+
+        if (ctrlKey || metaKey) {
+            this.#list.selectTask(textContent);
+            this.#list.clearSingleClickIndex();
+        } else if (shiftKey) {
+            this.#list.selectRange(textContent);
+        } else {
+            this.#list.selectSingleTask(textContent);
+        }
+
+        this.#render();
+    }
+
+    #render() {
```

Comment:
It might be resource-intensive. I'd advise finding a way to avoid rendering when it is possible